src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,224 @@
+    }
+    switch (params_.img_frame) {
+    case image_frame::file:
+        return make_intrusive<blob_record_reader>(ds.open_read());
+    case image_frame::pipe:
+        return make_intrusive<detail::recordio_record_reader>(ds.open_read());
+    }
+}
+
+void
+image_reader::infer_schema(instance const &)
+{
+    std::vector<feature_desc> descs{};
+    // schema follows the NHWC convention
+    descs.emplace_back(
+        feature_desc_builder{"value",
+                             data_type::uint8,
+                             {params().batch_size,
+                              static_cast<std::size_t>(params_.data_shape[1]),
+                              static_cast<std::size_t>(params_.data_shape[2]),
+                              static_cast<std::size_t>(params_.data_shape[0])}}
+            .build());
+    schema_ = make_intrusive<schema>(std::move(descs));
+}
+
+intrusive_ptr<example>
+image_reader::decode(instance_batch const &batch) const
+{
+    auto tsr = make_tensor(batch.instances(), batch.size());
+    std::vector<intrusive_ptr<tensor>> tensors{};
+    tensors.emplace_back(tsr);
+    return make_intrusive<example>(schema_, std::move(tensors));
+}
+
+intrusive_ptr<dense_tensor>
+image_reader::make_tensor(std::vector<instance> const &instances,
+                          std::size_t batch_size) const
+{
+    auto instance_offset =
+        params_.data_shape[0] * params_.data_shape[1] * params_.data_shape[2];
+    auto total_memory_buffer =
+        static_cast<unsigned long>(instance_offset) * batch_size;
+    std::vector<std::uint8_t> img_vec(total_memory_buffer);
+    auto *ptr_img_vec = img_vec.data();
+    int type;
+    cv::ImreadModes mode;
+    for (const instance &ins : instances) {
+        memory_slice img_buf;
+        if (params_.img_frame == image_frame::pipe) {
+            img_buf = ins.bits().subslice(recordio_image_header_offset_);
+        }
+        else {
+            img_buf = ins.bits();
+        }
+        cv::Mat mat(
+            1,
+            static_cast<int>(img_buf.size()),
+            CV_8U,
+            static_cast<void *>(const_cast<std::byte *>(img_buf.data())));
+        switch (params_.data_shape[0]) {
+        case 1:
+            mode = cv::ImreadModes::IMREAD_GRAYSCALE;
+            type = CV_8UC1;
+            break;
+        case 3:
+            mode = cv::ImreadModes::IMREAD_COLOR;
+            type = CV_8UC3;
+            break;
+        case 4:
+            mode = cv::ImreadModes::IMREAD_UNCHANGED;
+            type = CV_8UC4;
+            break;
+        default:
+            throw std::invalid_argument{
+                fmt::format("Unsupported number of channels: {0:n}",
+                            params_.data_shape[0])};
+        }
+        cv::Mat tmp = decode_image(mat, mode, ins);
+        if (params_.resize != -1) {
+            resize(tmp, tmp);
+        }
+        if (params_.to_rgb && tmp.channels() != 1) {
+            cv::cvtColor(tmp, tmp, cv::COLOR_BGR2RGB);
```

Comment:
I assume this operation can potentially fail. You should check its error status.

Reply:
method doesn't return an error code: https://docs.opencv.org/4.2.0/db/d64/tutorial_js_colorspaces.html

Reply:
As discussed offline, OpenCV uses exceptions for error propagation.